src/client_side_reply.cc

```diff
@@ -440,12 +440,18 @@ clientReplyContext::handleIMSReply(const StoreIOBuffer result)
 
     // origin replied 304
     if (status == Http::scNotModified) {
-        http->updateLoggingTags(LOG_TCP_REFRESH_UNMODIFIED);
-        http->request->flags.staleIfHit = false; // old_entry is no longer stale
-
         // TODO: The update may not be instantaneous. Should we wait for its
         // completion to avoid spawning too much client-disassociated work?
-        Store::Root().updateOnNotModified(old_entry, *http->storeEntry());
+        if (!Store::Root().updateOnNotModified(old_entry, *http->storeEntry())) {
+            old_entry->release(true);
+            restoreState();
+            http->updateLoggingTags(LOG_TCP_MISS);
+            processMiss();
+            return;
+        }
+
+        http->updateLoggingTags(LOG_TCP_REFRESH_UNMODIFIED);
+        http->request->flags.staleIfHit = false; // old_entry is no longer stale
```

Comment:
The diff does not show this well, but these two "new" code lines is the _old_ 304 handling code. No changes on this "successful 304 update" code path.